Common/include/toolboxes/ndflattener.hpp

```diff
@@ -0,0 +1,846 @@
+  };
+
+  /*! \class IndexAccumulator
+   * \brief Data structure holding an offset for the NdFlattener, to provide a []...[]-interface.
+   * \details Derived from IndexAccumulator_Base, specifying the operator[] method:
+   *  - For N==1, the structure has already read all indices but one. So after this method has read the last
+   *    index, return a reference to the data.
+   *  - The case N==2 is much like the case N>3 but an additional function data() should be provided.
+   *  - For N>3, more indices have to be read, return an IndexAccumulator<N-1>.
+   * \tparam N - Number of missing parameters
+   * \tparam Nd_type - Type of the accessed NdFlattener, should be NdFlattener<N,...>
+   */
+  /*! \class IndexAccumulator_Base
+   * \brief Parent class of IndexAccumulator.
+   * \details IndexAccumulator provides the operator[] method.
+   */
+  template<size_t _N, typename _Nd_type>
+  class IndexAccumulator_Base {
+  public:
+    static constexpr size_t N = _N;
+    using Nd_type = _Nd_type;
+
+    /*== Conditional is part of the standard library since C++14, keep it here for C++11 compatibility. ==*/
+    template<bool B, class T, class F> struct conditional { typedef T type; };
+    template<class T, class F> struct conditional<false, T, F> { typedef F type; };
+
+  protected:
+    Nd_type* nd; /*!< \brief The accessed NdFlattener. */
+    const typename Nd_type::Index offset; /*!< \brief Index in the currently accessed layer. */
+    //const typename Nd_type::Index size; /*!< \brief Exclusive upper bound for the next index. */
+    IndexAccumulator_Base(Nd_type* nd, typename Nd_type::Index offset):
+      nd(nd), offset(offset) {}
+
+  };
+
+  template<size_t _N, typename _Nd_type>
+  class IndexAccumulator : public IndexAccumulator_Base<_N,_Nd_type>{
+  public:
+    static constexpr size_t N = _N;
+    using Nd_type = _Nd_type;
+    typedef IndexAccumulator_Base<_N,_Nd_type> Base;
+
+    template<class ...ARGS> IndexAccumulator(ARGS... args): Base(args...) {}
+
+    /*! The Base of NdFlattener<K> is NdFlattener<K-1>, but do also preserve constness.
+     */
+    using Nd_Base_type = typename Base::template conditional<
+      std::is_const<Nd_type>::value,
+      const typename Nd_type::Base,
+      typename Nd_type::Base
+    >::type;
+    /*! Return type of operator[]. */
+    using LookupType = IndexAccumulator<N-1, Nd_Base_type>;
+
+    /*! \brief Read one more index.
+     * \param[in] i - Index.
+     */
+    LookupType operator[] (typename Nd_type::Index i) {
+      //const typename Nd_type::Index new_offset = this->nd->GetIndices()[this->offset+i];
+      //const typename Nd_type::Index new_size = this->nd->GetIndices()[this->offset+i+1] - new_offset;
+      return LookupType(static_cast<Nd_Base_type*>(this->nd),this->nd->GetIndices()[this->offset+i]);
+    }
+  };
+
+  template<size_t _N, typename _Nd_type>
+  class IndexAccumulator_Checked : public IndexAccumulator<_N,_Nd_type>{
+  public:
+    static constexpr size_t N = _N;
```

Comment:
I understand the motivation, but this duplication is probably not justified.
Add size to the "normal IndexAccumulator" since that would be useful to allow range-based loops.
Always assert(i<=size) (since it goes away in optimized builds).
Add a bool template param (i.e. Check) and if Check is true you do the MPI:Error thing.

Reply:
Ok, I "merged" the IndexAccumulator_Checked into IndexAccumulator and IndexAccumulator_Base